feat(aws/cloudformation): Add task to deploy a cloudformation stack

xavileon opened the pull request:
This task is responsible to deploy a cloudformation stack. It receives a
description of the cloudformation stack containing:
* the name of the stack
* the region where it's deployed
* the template of the cloudformation stack
* and the parameters used for the template

This will create the stack process in AWS and return the stackId as a
result of the task, so Orca can followup on its progress.

This PR partly addresses spinnaker/spinnaker#736 and depends on https://github.com/spinnaker/clouddriver/pull/3278

Signed-off-by: Xavi León <xavi.leon@schibsted.com>

gardleopard commented:
@spinnaker/amazon PTAL

xavileon commented:
@spinnaker/reviewers PTAL

xavileon commented:
@spinnaker/reviewers PTAL

robzienert commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java:
```diff
@@ -0,0 +1,115 @@
+  AmazonClientProvider amazonClientProvider;
+
+  @Autowired
+  private ObjectMapper objectMapper;
```
I don't think this is going to get autowired? Are you sure it does?

xavileon replied:
Yes, it gets autowired 👍 configured in https://github.com/spinnaker/clouddriver/blob/master/clouddriver-aws/src/main/groovy/com/netflix/spinnaker/config/AwsConfiguration.groovy#L122

robzienert commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java:
```diff
@@ -0,0 +1,115 @@
+    try {
+      template = objectMapper.writeValueAsString(description.getTemplateBody());
+    } catch (JsonProcessingException e) {
+      throw new IllegalArgumentException("Could not deserialize template body", e);
```
We're serializing here.

robzienert commented on clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/ops/DeployCloudFormationAtomicOperation.java:
```diff
@@ -0,0 +1,115 @@
+      .withStackName(description.getStackName())
+      .withParameters(parameters)
+      .withTemplateBody(template);
+    task.updateStatus(BASE_PHASE, "Uploading cloudformation");
```
I'd recommend changing these to say "[blah blah] CloudFormation Stack" instead.

xavileon replied:
good point, updated 👍

robzienert commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java:
```diff
@@ -94,4 +94,8 @@
   public static final String DESTROY_SERVICE = "destroyService";
+
+  // CloudFormation operations
+  public static final String DEPLOY_CLOUDFORMATION = "deployCloudFormation";
```
`DEPLOY_CLOUDFORMATION_STACK` imo.

xavileon commented:
@robzienert Thanks for looking into it 👍 Addressed your comments in https://github.com/spinnaker/clouddriver/pull/3280/commits/f6857c526c9659f5d10e4b817ae5a156fcd2891e

xavileon commented:
hey @robzienert addressed your feedback, is it ok to merge?

xavileon commented on clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/orchestration/AtomicOperations.java:
```diff
@@ -94,4 +94,8 @@
   public static final String DESTROY_SERVICE = "destroyService";
+
+  // CloudFormation operations
+  public static final String DEPLOY_CLOUDFORMATION_STACK = "deployCloudFormation";
```
Sorry, I'm still getting familiar with clouddriver 🙂 What do you mean? Could you point me to a task/operation that is not an AtomicOperation and could be an example to follow? If that suits our need I'm happy to make the necessary changes.

ajordens replied:
I can circle around on this post-merge ... let's get it down for now.
